olcut-core/src/main/java/com/oracle/labs/mlrg/olcut/config/ConfigurationData.java

```diff
@@ -263,6 +274,151 @@ public int hashCode() {
+            propNames.addAll(bRoot.getProperties().keySet());
+
+            boolean typesMatch = aRoot.getClassName().equals(bRoot.getClassName());
+            if(!typesMatch) {
+                logger.info(String.format("type mismatch, a.class: %s b.class: %s", aRoot.getClassName(), bRoot.getClassName()));
+            }
+
+            return typesMatch &&
+                    propNames.stream().allMatch(propName -> {
+                        Optional<Property> aPropOpt = Optional.ofNullable(aRoot.getProperties().get(propName));
+                        Optional<Property> bPropOpt = Optional.ofNullable(bRoot.getProperties().get(propName));
+                        if(aPropOpt.isPresent() && bPropOpt.isPresent()) {
+                            Property aProp = aPropOpt.get();
+                            Property bProp = bPropOpt.get();
+                            if(aProp instanceof SimpleProperty && bProp instanceof SimpleProperty) {
+                                return simplePropertyDerefEquals(a, b, (SimpleProperty) aProp, (SimpleProperty) bProp, propName);
+                            } else if(aProp instanceof ListProperty && bProp instanceof ListProperty) {
+                                List<SimpleProperty> aList = ((ListProperty) aProp).getSimpleList();
+                                List<SimpleProperty> bList = ((ListProperty) bProp).getSimpleList();
+                                return aList.size() == bList.size() &&
+                                        IntStream.range(0, aList.size()).allMatch(i -> {
+                                            boolean eq = simplePropertyDerefEquals(a, b, aList.get(i), bList.get(i), propName);
+                                            if(!eq) {
+                                                logger.info(String.format("Property key: %s, List index: %d, a.value: %s, b.value: %s",
+                                                        propName, i, aList.get(i).getValue(), bList.get(i).getValue()));
+                                            }
+                                            return eq;
+                                        });
+                            } else if(aProp instanceof MapProperty && bProp instanceof MapProperty) {
+                                Map<String, SimpleProperty> aMap = ((MapProperty) aProp).getMap();
+                                Map<String, SimpleProperty> bMap = ((MapProperty) bProp).getMap();
+                                return aMap.keySet().equals(bMap.keySet()) &&
+                                        aMap.keySet().stream().allMatch(k -> {
+                                            boolean eq = simplePropertyDerefEquals(a, b, aMap.get(k), bMap.get(k), propName);
+                                            if(!eq) {
+                                                logger.info(String.format("Property key: %s, PropValue Key: %s, a.value: %s, b.value: %s",
```

Comment:
Doesn't `simplePropertyDerefEquals` already log this? If so, then this lambda can drop the return statement and it'll be much nicer.